pydrofoil/real.py

```diff
@@ -0,0 +1,264 @@
+                # 1.222e+5
+                # 1.222222e+3
+            elif s[j] == "-":
+                shift = int(s[j+1:])
+                num = rbigint.fromstr(s[:i]+s[i+1:j-1])
+                dif = j-i-2+shift
+                if dif < 19:
+                    den = rbigint.fromint(10**dif)
+                else:
+                    den = den_of_fromstr(dif)
+                return Real(num, den)
+
+    # string to real without considering about input as x.xxxxe+xxx or x.xxxxe-xxxx
+    # @staticmethod
+    # def fromstr(str):
+    #     from rpython.rlib.rstring import strip_spaces
+    #     s = strip_spaces(str)
+    #     lead_zero_count = 0
+    #     ZERO_COUNT_END = False
+    #     for i in range(0, len(s)):
+    #         if s[i].isdigit() and s[i] != "0":
+    #             ZERO_COUNT_END = True
+    #         if (not ZERO_COUNT_END) and s[i] == "0":
+    #             lead_zero_count += 1 
+    #         if not s[i].isdigit() and s[i] != "+" and s[i] != "-":
+    #             for j in range(i+1, len(s)):
+    #                 if (not ZERO_COUNT_END) and s[j] == "0":
+    #                     lead_zero_count += 1
+    #                 else:
+    #                     break
+    #             break
+    #     # return i, s[:i]+s[i+1:]
+    #     if lead_zero_count == len(s)-1:
+    #         return Real(rbigint.fromint(0), rbigint.fromint(1))
+    #     num = rbigint.fromstr((s[:i] + s[i+1:])[lead_zero_count:] if (s[0] != "+" and s[0] != "-") else s[0]+(s[1:i] + s[i+1:])[lead_zero_count:]) if i < len(s)-1 else rbigint.fromstr(s)
+    #     dif = len(s)-1-i
+    #     if dif < 19:
+    #         den = rbigint.fromint(10**dif)
+    #     else:
+    #         den = den_of_fromstr(dif)
+    #     return Real(num, den)
+    #     # num = ((s[:i] + s[i+1:])[lead_zero_count:] if (s[0] != "+" and s[0] != "-") else s[0]+(s[1:i] + s[i+1:])[lead_zero_count:]) if i < len(s)-1 else (s)
+    #     # dif = len(s)-1-i
+    #     # if dif < 19:
+    #     #     den = rbigint.fromint(10**dif)
+    #     # else:
+    #     #     den = den_of_fromstr(dif)
+    #     # # return Real(num, den)
+    #     # return lead_zero_count, num, den.str()
+
+
+
+    def add(self, other):
+        num_new_1 = self.num.mul(other.den)
+        num_new_2 = other.num.mul(self.den)
+        den_new = self.den.mul(other.den)
+        num_new = num_new_1.add(num_new_2)
+        return Real(num_new, den_new)
+    
+    def sub(self, other):
+        num_new_1 = self.num.mul(other.den)
+        num_new_2 = other.num.mul(self.den)
+        den_new = self.den.mul(other.den)
+        num_new = num_new_1.sub(num_new_2)
+        return Real(num_new, den_new)
+    
+    def mul(self, other):
+        num_new = self.num.mul(other.num)
+        den_new = self.den.mul(other.den)
+        return Real(num_new, den_new)
+    
+    def div(self, other):
+        if self.num.int_eq(0) and not other.num.int_eq(0):
```

Comment:
is this case even needed? if `self.num` is 0, then the `else` case below will just do the right thing too, no?

Reply:
oh yes, I add this condition just because it can return the Real type directly without doing the multiplication of rbigint, but I'm not so sure whether it will improve the efficiency to some extent

Reply:
let's ignore efficiency for now. we will have to think about efficiency soon anyway, but I would like things to be simple and correct first.